dabudimesh/router.py

```diff
@@ -9,20 +11,20 @@ def __init__(self, address, listener):
 
     def add_connection(self, address, socket):
         # TODO: inform neighbors recursively
-        socket.sendall(bytes(self.address, "utf-8"))
+        for key in self.routing_table:
+            message = Message(self.address, "add_address", key, address)
+            self.send(message)
         self.routing_table[address] = socket
 
-    def send(self, address, message):
-        # TODO: raise an error if not found
-        socket = self.routing_table[address]
-        socket.sendall(bytes(f"{address}:{message}", "utf-8"))
+    def send(self, message):
+        socket = self.routing_table[message.fields["address"]]
+        socket.sendall(message.encode())
 
     def process_message_from(self, socket):
-        message = socket.recv(MSG_MAX_LEN).decode()
-        [address, text] = message.split(":", maxsplit=1)
-        if address == self.address:
-            print(f"Read: {text}")
-            return text
+        message = Message("0", "message", "0")
+        message.decode(socket.recv(MSG_MAX_LEN).decode("utf-8"))
+        if int(message.fields["destination"]) == self.address:
```

Comment:
Type of self.address is string, so do not cast message.fields["destination"] to int

Reply:
Fixed